[WIP] EIP-7594: PeerDAS protocol (#3574)
ethereum/consensus-specs: specs/_features/eip7594/das-core.md

    - [`get_data_column_sidecars`](#get_data_column_sidecars)
- [Custody](#custody)
  - [Custody requirement](#custody-requirement)
  - [Public, deterministic selection](#public-deterministic-selection)
- [Peer discovery](#peer-discovery)
- [Extended data](#extended-data)
- [Column gossip](#column-gossip)
  - [Parameters](#parameters)
  - [Reconstruction and cross-seeding](#reconstruction-and-cross-seeding)
- [Peer sampling](#peer-sampling)
- [Peer scoring](#peer-scoring)
- [DAS providers](#das-providers)
- [A note on fork choice](#a-note-on-fork-choice)
- [FAQs](#faqs)
  - [Row (blob) custody](#row-blob-custody)
  - [Subnet stability](#subnet-stability)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->
<!-- /TOC -->

## Custom types

We define the following Python custom types for type hinting and readability:

| Name | SSZ equivalent | Description |
| - | - | - |
| `DataColumn` | `List[Cell, MAX_BLOBS_PER_BLOCK]` | The data of each column in EIP-7594 |
| `ExtendedMatrix` | `List[Cell, MAX_BLOBS_PER_BLOCK * NUMBER_OF_COLUMNS]` | The full data of one-dimensional erasure coding extended blobs (in row major format) |

> | `FlatExtendedMatrix` | `List[BLSFieldElement, MAX_BLOBS_PER_BLOCK * FIELD_ELEMENTS_PER_BLOB * NUMBER_OF_COLUMNS]` | The flattened format of `ExtendedMatrix` |

Reviewer: ```suggestion
| `FlatExtendedMatrix` | `List[BLSFieldElement, FIELD_ELEMENTS_PER_CELL * MAX_BLOBS_PER_BLOCK * NUMBER_OF_COLUMNS]` | The flattened format of `ExtendedMatrix` |
```

As discussed in https://github.com/hwwhww/consensus-specs/pull/1#discussion_r1417574161

## Configuration

### Data size

| Name | Value | Description |
| - | - | - |
| `NUMBER_OF_COLUMNS` | `uint64((FIELD_ELEMENTS_PER_BLOB * 2) // FIELD_ELEMENTS_PER_CELL)` (= 128) | Number of columns in the extended data matrix. |

### Custody setting

| Name | Value | Description |
| - | - | - |

> | `SAMPLES_PER_SLOT` | `8` | Number of `DataColumn` random samples a node queries per slot |

Reviewer 2: Does it make sense to define a config/preset value for the probability of the data being available/unavailable and derive `SAMPLES_PER_SLOT` from that number? I'm not clear on how we arrived at this number.

Reviewer 3: yes this should be related with probablity, however i would prefer to have this as the config/preset constant, probablity analysis vs sample size could be attached with the specs

Reviewer: > Does it make sense to define a config/preset value for the probability of the data being available/unavailable and derive `SAMPLES_PER_SLOT` from that number? I'm not clear on how we arrived at this number.

If >50% of the columns are missing, the probability that `is_data_available` is true is simply at most `0.5^(SAMPLE_PER_SLOT)`

We have `8` because there would be expectedly at most 1/256 of the validators that have `is_data_available` to be true when it should be false.

Reviewer 4: I don't think that `SAMPLES_PER_SLOT` should be as low as 8. Sampling isn't the bandwidth bottleneck (distribution through subnets is), and 8 samples leaves a very-far-from-negligible probability of a full node experiencing a safety violation (chain looks available and confirmed, but it's not), *even without a targeted attack*. 
I would advocate for `SAMPLES_PER_SLOT = 16`.

Reviewer 5: Just brainstorming about resiliency without adding significant security risk, some alternatives came to mind:
- Is there any downside of collecting samples from a fixed number of peers? (e.g. getting 16 samples from 8 peers, instead of potentially up to 16 peers) I think the current proposal is more flexible (in terms of param tweaking) and covers sampling against more peers than this, however I think there's a tradeoff on robustness, i.e. successfully getting all samples from all 16 peers could be harder than getting from 8 peers.
- Is there any security risk if we randomly select 8 connected peers that are assigned different columns, rather than selecting 8 columns then look for the peers if we don't already connect to them?

Reviewer 4: > * Is there any downside of collecting samples from a fixed number of peers? (e.g. getting 16 samples from 8 peers, instead of potentially up to 16 peers) I think the current proposal is more flexible (in terms of param tweaking) and covers sampling against more peers than this, however I think there's a tradeoff on robustness, i.e. successfully getting all samples from all 16 peers could be harder than getting from 8 peers.

It depends what kind of security you are concerned with. From the network's perspective, it makes no difference how you get your samples, it is only important that you do. Everyone could get them from a single source and it wouldn't make any difference. 

From an individual node's perspective, it does matter somewhat, because getting them from multiple nodes (assuming they are actually unrelated) would make you less susceptible to targeted release of data. Without getting into details too much, I don't think this is something to be particularly concerned about, and I don't think there is much of a reason to try hard to get each sample from a different peer.

> Is there any security risk if we randomly select 8 connected peers that are assigned different columns, rather than selecting 8 columns then look for the peers if we don't already connect to them?

Yes, there is. The columns you choose *must* be random. You should not treat the node ids of your peers as random.

| `CUSTODY_REQUIREMENT` | `2` | Minimum number of columns an honest node custodies and serves samples from |
| `TARGET_NUMBER_OF_PEERS` | `70` | Suggested minimum peer count |

### Helper functions

#### `get_custody_lines`

```python

> def get_custody_lines(node_id: NodeID, custody_size: uint64) -> Sequence[ColumnIndex]:

Reviewer 3: yes it could be a function of EPOCH/SOME_CONSTANT and node_id

A node stores the custodied columns for the duration of the pruning period and responds to peer requests for samples on those columns.

### Public, deterministic selection

The particular columns that a node custodies are selected pseudo-randomly as a function (`get_custody_lines`) of the node-id and custody size -- importantly this function can be run by any party as the inputs are all public.

*Note*: increasing the `custody_size` parameter for a given `node_id` extends the returned list (rather than being an entirely new shuffle) such that if `custody_size` is unknown, the default `CUSTODY_REQUIREMENT` will be correct for a subset of the node's custody.

## Peer discovery

At each slot, a node needs to be able to readily sample from *any* set of columns. To this end, a node should find and maintain a set of diverse and reliable peers that can regularly satisfy their sampling demands.

> A node runs a background peer discovery process, maintaining at least `TARGET_NUMBER_OF_PEERS` of various custody distributions (both custody_size and column assignments). The combination of advertised `custody_size` size and public node-id make this readily and publicly accessible.

Reviewer: We cannot assume that the node ids are random. Joining the discv5 DHT is permissionless, so the attacker can flood the DHT with their own distribution of node ids. In addition, we don't always assume that we will always do the random lookup even though we are doing right now.

## Extended data

In this construction, we extend the blobs using a one-dimensional erasure coding extension. The matrix comprises maximum `MAX_BLOBS_PER_BLOCK` rows and fixed `NUMBER_OF_COLUMNS` columns, with each row containing a `Blob` and its corresponding extension.

## Column gossip

### Parameters

For each column -- use `data_column_sidecar_{subnet_id}` subnets, where each column index maps to the `subnet_id`. The sidecars can be computed with `get_data_column_sidecars(signed_block: SignedBeaconBlock, blobs: Sequence[Blob])` helper.

To custody a particular column, a node joins the respective gossip subnet. Verifiable samples from their respective column are gossiped on the assigned subnet.

### Reconstruction and cross-seeding

If the node obtains 50%+ of all the columns, they can reconstruct the full data matrix via `recover_samples_impl` helper.

If a node fails to sample a peer or fails to get a column on the column subnet, a node can utilize the Req/Resp message to query the missing column from other peers.

> Once the node obtain the column, the node should send the missing columns to the column subnets.

Reviewer 6: > the node should send the missing columns to the column subnets.

Can we clarify what "should send" means here? I see a few interpretations:
- The node is assigned to a column subnet, but nothing shows up. In this case, the node reconstructs the column and gossip back to the peers on the same column subnet the column blob sidecar
- The node is having trouble with everything because it's only gotten 49% of the columns, it manually requests the last column that made it to 50%, and the node gossip back that last missing 1% column back to the subnet. Not original assigned

Reviewer 3: I think, if node doesn't get a particular `DataColumn` that it needs to sample, it should then attempt to reconstruct the blob using req/resp and then seed it to the network.

with regard to second point, a node will ideally not be requiring 50% columns download unless it chooses to because then there are no actual sampling (if i understand you correctly)

Reviewer: > The node is having trouble with everything because it's only gotten 49% of the columns, it manually requests the last column that made it to 50%, and the node gossip back that last missing 1% column back to the subnet. Not original assigned

The node is not supposed to download that many columns as @g11tech said, unless it wants to reconstruct.

> The node is assigned to a column subnet, but nothing shows up. In this case, the node reconstructs the column and gossip back to the peers on the same column subnet the column blob sidecar

Afaiu, we don't have a criteria yet when the node should consider the column as missing.

> I think, if node doesn't get a particular `DataColumn` that it needs to sample, it should then attempt to reconstruct the blob using req/resp and then seed it to the network (that could be the `robustness` bit of the peerDAS)

We also haven't decided yet who will do the reconstruction, since the reconstruction is quite an expensive job.

Reviewer 3: I think for e.g. data availability needs to be verified for the parent block before current block is imported, so the node needs to draw the conclusion before the block import else it can't do the import.

imo the nodes doing the `attestation`/proposal/aggregation would definitely need to do reconstruction else they won't be able to perform their dutites regarding a head that is not available in their forkchoice, but ideally any node could/should do reconstruction and then seed data in their subnets.

so essentially all validator connected nodes should definitely do it as a preparation step for next slot

Reviewer 5: @g11tech would you mind explaining why nodes need to reconstruct the full matrix to perform validator duties?

Reviewer 4: In the current spec, a normal node with `custody_size = 1` cannot help with reconstruction. If a block is > 50% available but not fully available, the assumption is that there is at least one node downloading everything (with `custody_size = 32`, participating in every subnet) which will do reconstruction and re-seed the missing columns *before the next slot*, so that everyone's sampling succeeds and they can import the next proposed block and vote for it

Reviewer 3: > @g11tech would you mind explaining why nodes need to reconstruct the full matrix to perform validator duties?

If sampling fails then availability fails, how can then one import the head block in the forkchoice? so there are two mitigations:

1. Reconstruct by req/rep with root/index and import if one can reconstruct (and also seed the failed sampled but constructed column to its subnet)
2. Or not import the block at all to the forkchoice

Reviewer 5: What does eventually succeed mean? Having retrieved 8 valid samples successfully and _consecutively_ before the DA filter cutoff time?

If it's not consecutive successes, we could eventually get 8 samples even if there are only 8 (out of 128) of them available right?

Reviewer 5: *consecutive* successes is probably too tight and not necessary (given we retry on _the same column_) , as this also relies on the network conditions and the peers.

I think my confusion came from "not retrying on the same column" (randomly pick another column), which doesn't make sense at all 😅 

So I think this is what it means :
- A node has randomly selected to sample 8 columns
- It would try getting samples from those 8 columns
- if it fails to get 1 column from a peer, it would try to retrieve the _same missing column_ with another peer until it finds one
- if the node successfully retrieve samles for the randomly selected 8 columns, DA check is successful, otherwise it fails the DA check.

Please correct me if I'm wrong 🙏

Reviewer 4: Yes, exactly! It's important that you don't just change your choice of columns to sample because of failing. Either you get all of the columns for the random indices you originally chose, or sampling fails

Reviewer 5: Just had a chat with @AgeManning and think we might need some tolerance here with sampling as it's likely that we won't have time to "find peers" if there aren't any peers in a randomly selected column. 

We should have a background task that tries to maintain a uniform set across all columns, but there would still times where we do not have any peers in a given column, and it might be too late to look for a peer as discovery is slow, and even once we find a peer, there's no guarantee that we can connect to them (could be dead, reached peer limit etc). 

Ideally we have some tolerance on sampling here, i.e. rather than requiring 16 out of 16 selected columns to pass, could we require 12 out of 16? (just some make up numbers, could be a higher percentage, but perhaps not 100%)

Reviewer 7: @jimmygchen  @fradamt, to follow up on this, now with the PeerDAS specific numbers:

if you set a false positive threshold of 1e-3, you would need 10 columns if losses are not permitted. Or, you can go for 13 columns, with up to 1 loss. 16 columns with up to 2.
Using incrementalDAS, you can do this as follows:
- you start by doing a random permutation P of the columnIDs. P has 128 elements.
- you request the first 10 columns of the random permutation P[0..9] with a given timeout, say 3 seconds.
- if by 3 seconds all are there, test passed, at minimal network cost.
- if 1 missing, request the missing column from another peer, AND request P[10-12] from peers. All with a new, shorter timeout. Hope to get enough (12 out if 13) by the new timeout.
- and so on ....

Obviously, we can easily adapt this to a lower (more stringent) FP threshold to start with 16 columns.

LossyDAS for PeerDAS calculations: https://colab.research.google.com/drive/18uUgT2i-m3CbzQ5TyP9XFKqTn1DImUJD

*Note*: A node always maintains a matrix view of the rows and columns they are following, able to cross-reference and cross-seed in either direction.

*Note*: There are timing considerations to analyze -- at what point does a node consider samples missing and choose to reconstruct and cross-seed.

*Note*: There may be anti-DoS and quality-of-service considerations around how to send samples and consider samples -- is each individual sample a message or are they sent in aggregate forms.

## Peer sampling

> At each slot, a node makes (locally randomly determined) `SAMPLES_PER_SLOT` queries for samples from their peers via `DataColumnSidecarByRoot` request. A node utilizes `get_custody_lines` helper to determine which peer(s) to request from. If a node has enough good/honest peers across all rows and columns, this has a high chance of success.

Reviewer 8: Is there more information on how a peer determines for which column they would randomly sample ? Or do we just use a local source of RNG and use that to choose which column samples we want

Reviewer: > Or do we just use a local source of RNG and use that to choose which column samples we want

Yes, currently that's right.

## Peer scoring

> Due to the deterministic custody functions, a node knows exactly what a peer should be able to respond to. In the event that a peer does not respond to samples of their custodied rows/columns, a node may downscore or disconnect from a peer.

Reviewer 2: I think it would be way better implementation wise if the spec defines a particular time within the slot when we should do the sampling. 
Otherwise, there's no distinction between the scenario where a peer who is supposed to custody column `X` 
1. hasn't yet received the full sample from the block producer
2. has received but isn't serving

which would complicate peer scoring considerably implementation wise.

If the trailing fork choice filter described below is viable security wise, then maybe we only sample at the 4 second mark where validators have to attest to slot N

Reviewer 3: i think since it's for the past slot, so it should defintely be completed by before you import the block, one could try doing it in the last 3 seconds of previous block?

Reviewer: I think, timing is TBD.

Reviewer 4: Imo it would make sense to start sampling as soon as possible, e.g. as soon as you receive either the block or a sidecar (which has all of the commitments). I am not sure how this affect peer scoring. Perhaps a node could downscore a peer only if they fail to provide a sample after some deadline, and if the node has good  certainty that this sample should be available? E.g. for samples for slot n, you could wait until you receive attestations from slot n, and only downscore peers if the block is canonical

DAS providers can also be found out-of-band and configured into a node to connect to directly and prioritize. Nodes can add some set of these to their local configuration for persistent connection to bolster their DAS quality of service.

Such direct peering utilizes a feature supported out of the box today on all nodes and can complement (and reduce attackability and increase quality-of-service) alternative peer discovery mechanisms.

## A note on fork choice

*Fork choice spec TBD, but it will just be a replacement of `is_data_available()` call in Deneb with column sampling instead of full download. Note the `is_data_available(slot_N)` will likely do a `-1` follow distance so that you just need to check the availability of slot `N-1` for slot `N` (starting with the block proposer of `N`).*

The fork choice rule (essentially a DA filter) is *orthogonal to a given DAS design*, other than the efficiency of a particular design impacting it.

In any DAS design, there are probably a few degrees of freedom around timing, acceptability of short-term re-orgs, etc. 

> For example, the fork choice rule might require validators to do successful DAS on slot N to be able to include block of slot `N` in its fork choice. That's the tightest DA filter. But trailing filters are also probably acceptable, knowing that there might be some failures/short re-orgs but that they don't hurt the aggregate security. For example, the rule could be — DAS must be completed for slot N-1 for a child block in N to be included in the fork choice.

Reviewer 3: ```suggestion
For example, the fork choice rule might require validators to do successful DAS on the parent's slot to be able to include block of slot `N` in its fork choice. That's the tightest DA filter. But trailing filters are also probably acceptable, knowing that there might be some failures/short re-orgs but that they don't hurt the aggregate security. For example, the rule could be — DAS must be completed for slot N-1 for a child block in N to be included in the fork choice.
```

Reviewer 5: I think the example is correct for DAS in the same slot? (tightest DA filter)

A node runs a background peer discovery process, maintaining at least `TARGET_NUMBER_OF_PEERS` of various custody distributions (both custody_size and column assignments). The combination of advertised `custody_size` size and public node-id make this readily and publicly accessible.

`TARGET_NUMBER_OF_PEERS` should be tuned upward in the event of failed sampling.

*Note*: while high-capacity and super-full nodes are high value with respect to satisfying sampling requirements, a node should maintain a distribution across node capacities as to not centralize the p2p graph too much (in the extreme becomes hub/spoke) and to distribute sampling load better across all nodes.

*Note*: A DHT-based peer discovery mechanism is expected to be utilized in the above. The beacon-chain network currently utilizes discv5 in a similar method as described for finding peers of particular distributions of attestation subnets. Additional peer discovery methods are valuable to integrate (e.g., latent peer discovery via libp2p gossipsub) to add a defense in breadth against one of the discovery methods being attacked.

## Extended data

> In this construction, we extend the blobs using a one-dimensional erasure coding extension. The matrix comprises maximum `MAX_BLOBS_PER_BLOCK` rows and fixed `NUMBER_OF_COLUMNS` columns, with each row containing a `Blob` and its corresponding extension.

Reviewer: How often will `ExtendedBlob` be used? I guess the new type will be rarely used in the code. I'm not sure. I may be wrong.

> In this construction, we extend the blobs using a one-dimensional erasure coding extension. The matrix comprises maximum `MAX_BLOBS_PER_BLOCK` rows and fixed `NUMBER_OF_COLUMNS` columns, with each row containing a `Blob` and its corresponding extension.

Reviewer 9: I imagine `ExtendedBlob` would be used as an intermediate type before converting the data into columns. So it would be used when publishing and reconstructing. You're right though, it might not be very useful. I don't have a strong opinion about having this type or not.

## FAQs

### Row (blob) custody

In the one-dimension construction, a node samples the peers by requesting the whole `DataColumn`. In reconstruction, a node can reconstruct all the blobs by 50% of the columns. Note that nodes can still download the row via `blob_sidecar_{subnet_id}` subnets.

The potential benefits of having row custody could include:

1. Allow for more "natural" distribution of data to consumers -- e.g., roll-ups -- but honestly, they won't know a priori which row their blob is going to be included in in the block, so they would either need to listen to all rows or download a particular row after seeing the block. The former looks just like listening to column [0, N)  and the latter is req/resp instead of gossiping.
2. Help with some sort of distributed reconstruction. Those with full rows can compute extensions and seed missing samples to the network. This would either need to be able to send individual points on the gossip or would need some sort of req/resp faculty, potentially similar to an `IHAVEPOINTBITFIELD` and `IWANTSAMPLE`.

However, for simplicity, we don't assign row custody assignments to nodes in the current design.

> 
> 
> ### Subnet stability

Reviewer 9: ```suggestion

### Subnet stability
```

    - [`get_custody_lines`](#get_custody_lines)
    - [`compute_extended_data`](#compute_extended_data)
    - [`compute_extended_matrix`](#compute_extended_matrix)
    - [`get_data_column_sidecars`](#get_data_column_sidecars)
- [Custody](#custody)
  - [Custody requirement](#custody-requirement)
  - [Public, deterministic selection](#public-deterministic-selection)
- [Peer discovery](#peer-discovery)
- [Extended data](#extended-data)
- [Column gossip](#column-gossip)
  - [Parameters](#parameters)
  - [Reconstruction and cross-seeding](#reconstruction-and-cross-seeding)
- [Peer sampling](#peer-sampling)
- [Peer scoring](#peer-scoring)
- [DAS providers](#das-providers)
- [A note on fork choice](#a-note-on-fork-choice)
- [FAQs](#faqs)
  - [Row (blob) custody](#row-blob-custody)
  - [Subnet stability](#subnet-stability)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->
<!-- /TOC -->

## Custom types

We define the following Python custom types for type hinting and readability:

| Name | SSZ equivalent | Description |
| - | - | - |
| `DataColumn`   | `List[Cell, MAX_BLOBS_PER_BLOCK]` | The data of each column in EIP7594 |

> | `ExtendedMatrix` | `List[Cell, MAX_BLOBS_PER_BLOCK * NUMBER_OF_COLUMNS]` | The full data with blobs and one-dimensional erasure coding extension |

Reviewer 3: ```suggestion
| `ExtendedMatrix` | `List[Cell, MAX_BLOBS_PER_BLOCK * NUMBER_OF_COLUMNS]` | The full data with blobs and one-dimensional erasure coding extension in column major format|
```

Reviewer 3: or most likely its row major? i.e. full extended blob cells are sequentially stored

| `FlatExtendedMatrix` | `List[BLSFieldElement, MAX_BLOBS_PER_BLOCK * FIELD_ELEMENTS_PER_BLOB * NUMBER_OF_COLUMNS]` | The flattened format of `ExtendedMatrix` |

> | `LineIndex`    | `uint64` | The index of the rows or columns in `FlatExtendedMatrix` matrix |

Reviewer 3: ```suggestion
| `LineIndex`    | `uint64` | The index of the `BLSFieldElement`  in `FlatExtendedMatrix` matrix |
```
?

## Configuration

### Data size

| Name | Value | Description |
| - | - | - |
| `NUMBER_OF_COLUMNS` | `uint64((FIELD_ELEMENTS_PER_BLOB * 2) // FIELD_ELEMENTS_PER_CELL)` (= 128) | Number of columns in the extended data matrix. |

### Custody setting

| Name | Value | Description |
| - | - | - |

> | `SAMPLES_PER_SLOT` | `8` | Number of random samples a node queries per slot |

Reviewer 3: this is per blob or per block? (i am sure answer will be available below, but clarity here would also be nice)

    block = signed_block.message
    kzg_commitments_inclusion_proof = compute_merkle_proof(
        block.body,
        get_generalized_index(BeaconBlockBody, 'blob_kzg_commitments'),
    )
    cells_and_proofs = [compute_cells_and_proofs(blob) for blob in blobs]
    blob_count = len(blobs)
    cells = [cells_and_proofs[i][0] for i in range(blob_count)]
    proofs = [cells_and_proofs[i][1] for i in range(blob_count)]
    sidecars = []
    for column_index in range(NUMBER_OF_COLUMNS):
        column = DataColumn([cells[row_index][column_index]
                             for row_index in range(blob_count)])
        kzg_proof_of_column = [proofs[row_index][column_index]
                               for row_index in range(blob_count)]
        sidecars.append(DataColumnSidecar(
            index=column_index,
            column=column,
            kzg_commitments=block.body.blob_kzg_commitments,
            kzg_proofs=kzg_proof_of_column,
            signed_block_header=signed_block_header,
            kzg_commitments_inclusion_proof=kzg_commitments_inclusion_proof,
        ))
    return sidecars
```

## Custody

### Custody requirement

> Each node downloads and custodies a minimum of `CUSTODY_REQUIREMENT` columns per slot. The particular columns that the node is required to custody are selected pseudo-randomly (more on this below).

Reviewer 3: ```suggestion
Each node downloads and custodies a minimum of `CUSTODY_REQUIREMENT` consecutive columns per slot. The particular set of columns that the node is required to custody are selected pseudo-randomly (more on this below).
```

Author: Resolved by changing the custody assignment calculation (https://github.com/ethereum/consensus-specs/pull/3574#issuecomment-1914298172).

A node *may* choose to custody and serve more than the minimum honesty requirement. Such a node explicitly advertises a number greater than `CUSTODY_REQUIREMENT`  via the peer discovery mechanism -- for example, in their ENR (e.g. `custody_lines: 8` if the node custodies `8` columns each slot) -- up to a `NUMBER_OF_COLUMNS` (i.e. a super-full node).

A node stores the custodied columns for the duration of the pruning period and responds to peer requests for samples on those columns.

### Public, deterministic selection

> The particular columns that a node custodies are selected pseudo-randomly as a function (`get_custody_lines`) of the node-id and custody size -- importantly this function can be run by any party as the inputs are all public.

Reviewer 3: ```suggestion
The particular set of consecutive columns that a node custodies is selected pseudo-randomly as a function (`get_custody_lines`) of the node-id and custody size -- importantly this function can be run by any party as the inputs are all public.
```

Author: Resolved by changing the custody assignment calculation (https://github.com/ethereum/consensus-specs/pull/3574#issuecomment-1914298172).

To custody a particular column, a node joins the respective gossip subnet. Verifiable samples from their respective column are gossiped on the assigned subnet.

### Reconstruction and cross-seeding

If the node obtains 50%+ of all the columns, they can reconstruct the full data matrix via `recover_samples_impl` helper.

If a node fails to sample a peer or fails to get a column on the column subnet, a node can utilize the Req/Resp message to query the missing column from other peers.

Once the node obtain the column, the node should send the missing columns to the column subnets.

*Note*: A node always maintains a matrix view of the rows and columns they are following, able to cross-reference and cross-seed in either direction.

*Note*: There are timing considerations to analyze -- at what point does a node consider samples missing and choose to reconstruct and cross-seed.

*Note*: There may be anti-DoS and quality-of-service considerations around how to send samples and consider samples -- is each individual sample a message or are they sent in aggregate forms.

## Peer sampling

> At each slot, a node makes (locally randomly determined) `SAMPLES_PER_SLOT` queries for samples from their peers via `DataColumnSidecarByRoot` request. A node utilizes `get_custody_lines` helper to determine which peer(s) to request from. If a node has enough good/honest peers across all rows and columns, this has a high chance of success.

Reviewer 3: ```suggestion
At each slot, a node makes (locally randomly determined) `SAMPLES_PER_SLOT` queries for samples from their peers via `DataColumnSidecarByIndex` request. A node utilizes `get_custody_lines` helper to determine which peer(s) to request from. If a node has enough good/honest peers across all rows and columns, this has a high chance of success.
```
??

> At each slot, a node makes (locally randomly determined) `SAMPLES_PER_SLOT` queries for samples from their peers via `DataColumnSidecarByRoot` request. A node utilizes `get_custody_lines` helper to determine which peer(s) to request from. If a node has enough good/honest peers across all rows and columns, this has a high chance of success.

Reviewer 3: ```suggestion
At each slot, a node makes (locally randomly determined) `SAMPLES_PER_SLOT` queries for samples from their peers via `DataColumnSidecarByRoot` request. A node utilizes `get_custody_lines` helper to determine which peer(s) to request from. If a node has enough good/honest peers across all rows and columns, this has a high chance of success.
```
from all peers on a particular line or from a randomly sampled subset? (if more than 1 peer available)

Reviewer 10: similar to sync, etc, requesting from peers is a local load-balancing and other optimization. I don't think the spec should dictate

I need to get some random set of columns, I look at my peer-set who is distributed across a diverse set of column custody, I make the requests to the peers that are supposed to have what I need based on local strategy/hueristic

Reviewer 3: understood, but expecting nodes to `randomly` sample peers (from the column subnet) for lets say each column would put stronger guarantees on the DA proofs? else there is a chance to endup on  a hub/spoke model if nodes hook onto the super nodes.

Reviewer 10: The failure here ends up being network liveness rather than safety. If there is an over reliance on a hub and it fails.

There is certainly a chance, as peers attempt to balance diversity of peer with guarantees that they'll respond. I do think that it is a clients job to hueristically balance this to ensure that if the hub defaults, that they have adequate diversity to fill in the gaps.

this is similar to attnet peer balancing -- peers of high attnets are very valuable but an over-reliance can create a brittle network.

Fortunately, the failure of an overly hub-spoked network is a temporary liveness, rather than safety failure

Reviewer 7: With the current numbers we have, a given column is sampled often by a given node, on average every `NUMBER_OF_COLUMNS / SAMPLES_PER_SLOT` slots. That's a few minutes.

If nodes do some rotation or randomization of where to sample from, or e.g. ask two nodes for the same sample (maybe not for all samples, only where they miss liveness info on peers for the given column), they can easily keep a list of active responding peers for each column.

Speed matters, and we are already OK with duplicates (at least in the GossipSub part), and from the safety perspective it does not matter where the sample comes from. I would use this duplicate sampling strategy, but this is an implementation decision.

I think for the EIP a SHOULD sounds good. Something like `a node SHOULD maintain a diverse set of peers for each column by verifying responsiveness to sample queries` , without going into the details of random, round robin, duplicate, or something else.

Author: applied @cskiraly's suggestion in https://github.com/ethereum/consensus-specs/pull/3574/commits/8150f76e1c743644343371c9dab67004bd254fa7

thanks!

## Peer scoring

Due to the deterministic custody functions, a node knows exactly what a peer should be able to respond to. In the event that a peer does not respond to samples of their custodied rows/columns, a node may downscore or disconnect from a peer.

## DAS providers

A DAS provider is a consistently-available-for-DAS-queries, super-full (or high capacity) node. To the p2p, these look just like other nodes but with high advertised capacity, and they should generally be able to be latently found via normal discovery.

DAS providers can also be found out-of-band and configured into a node to connect to directly and prioritize. Nodes can add some set of these to their local configuration for persistent connection to bolster their DAS quality of service.

Such direct peering utilizes a feature supported out of the box today on all nodes and can complement (and reduce attackability and increase quality-of-service) alternative peer discovery mechanisms.

## A note on fork choice

> *Fork choice spec TBD, but it will just be a replacement of `is_data_available()` call in Deneb with column sampling instead of full download. Note the `is_data_available(slot_N)` will likely do a `-1` follow distance so that you just need to check the availability of slot `N-1` for slot `N` (starting with the block proposer of `N`).*

Reviewer 3: also worth thinking if this could impact SSF feature

Reviewer 10: This "buys us time" in the current non-SSF mode. If/when designing for SSF, we'd need to ensure that the sub-slots have sufficient time to perform whatever the DAS mode is then.

So agreed, but it's pretty tunable via slot/sub-slot time when redesigning for SSF

The fork choice rule (essentially a DA filter) is *orthogonal to a given DAS design*, other than the efficiency of a particular design impacting it.

In any DAS design, there are probably a few degrees of freedom around timing, acceptability of short-term re-orgs, etc. 

For example, the fork choice rule might require validators to do successful DAS on slot N to be able to include block of slot `N` in its fork choice. That's the tightest DA filter. But trailing filters are also probably acceptable, knowing that there might be some failures/short re-orgs but that they don't hurt the aggregate security. For example, the rule could be — DAS must be completed for slot N-1 for a child block in N to be included in the fork choice.

Such trailing techniques and their analysis will be valuable for any DAS construction. The question is — can you relax how quickly you need to do DA and in the worst case not confirm unavailable data via attestations/finality, and what impact does it have on short-term re-orgs and fast confirmation rules.

## FAQs

### Row (blob) custody

> In the one-dimension construction, a node samples the peers by requesting the whole `DataColumn`. In reconstruction, a node can reconstruct all the blobs by 50% of the columns. Note that nodes can still download the row via `blob_sidecar_{subnet_id}` subnets.

Reviewer 3: ```suggestion
In the one-dimension construction, a node samples the peers by requesting the whole `DataColumn`. In reconstruction, a node can reconstruct all the blobs by 50% of the columns. Note that nodes can still download the row via `blob_sidecar_{subnet_id}` subnets from advertised supernodes.
```

Reviewer 10: They can download these from any honest peers, as all honest peers have a deterministic custody requirement. 

Nodes that choose to custody beyond the baseline are a nice-to-have but are not requisite here

Reviewer 5: It seems unclear to me whether we deprecate the `blob_sidecar_{subnet_id}` topics initially - I feel like the original intention in the ethresearch [post](https://ethresear.ch/t/from-4844-to-danksharding-a-path-to-scaling-ethereum-da/18046) to deprecate these topics was to not add extra bandwidth requirement on top of 4844 (prior to implementation of distributed reconstruction), while allowing an increase in max blob count.

I don't see `MAX_BLOBS_PER_BLOCK` increased here, feels like it make sense to increase this with PeerDAS? (otherwise we aren't really scaling)

However, if we increase `MAX_BLOBS_PER_BLOCK`, we are unlikely able to make regular nodes custody all rows / blbos, so we probably would need to either
1. Deprecate row subnets, and reconstruct blobs when required (expensive and potentially slow?)
2. Keep the row subnets, but make nodes custody a subset of all row subnets using a similar approach, e.g. `get_custody_row` .

Reviewer 5: Sorry I commented pre-maturely - I've just read this: 

> However, for simplicity, we don't assign row custody assignments to nodes in the current design.

So does that mean each node custody _all_ rows and 1 column by default under the current proposal?

Reviewer 4: > So does that mean each node custody _all_ rows and 1 column by default under the current proposal?

Each node custodies 1 subnet, which corresponds to 4 columns, and it does not custody any row.

Reviewer 3: > Each node custodies 1 subnet, which corresponds to 4 columns, and it does not custody any row.
yes this is my understanding of it

Reviewer 5: > Each node custodies 1 subnet, which corresponds to 4 columns, and it does not custody any row.

How do we easily serve Blobs to users of the Beacon endpoints? 

Without any nodes custodying row subnets, we'd need to retrieve 50% of the columns and reconstruct the full maxtrix right?

Reviewer 5: I've just saw the DAS c-kzg benchmarks, the time for reconstruction isn't actually that bad, but wonder how quick we could get 50% of the samples.

Reviewer: > Without any nodes custodying row subnets, we'd need to retrieve 50% of the columns and reconstruct the full maxtrix right?

Yes. Actually, there is no need to serve specific blobs.

Author: Deprecated `blob_sidecar_{subnet_id}
 ` in https://github.com/ethereum/consensus-specs/pull/3574/commits/d23452d02bb54e9bffc5547fde00cfd1537d6d42.

Feel free to reopen it if needed :)

The potential benefits of having row custody could include:

> 1. Allow for more "natural" distribution of data to consumers -- e.g., roll-ups -- but honestly, they won't know a priori which row their blob is going to be included in in the block, so they would either need to listen to all rows or download a particular row after seeing the block. The former looks just like listening to column [0, N)  and the latter is req/resp instead of gossiping.

Reviewer 3: ```suggestion
1. Allow for more "natural" distribution of data to consumers -- e.g., roll-ups -- but honestly, they won't know apriori which row their blob is going to be included in in the block, so they would either need to listen to all rows or download a particular row after seeing the block. The former looks just like listening to column [0, N)  and the latter is req/resp instead of gossiping.
```

Author: Latin v.s. English?
Isn't "apriori" usually for the specific algorithm?

Reviewer 3: i guess you may reject the suggestion :+1: , don't have the rights to resolve the comment

| `ExtendedMatrix` | `List[Cell, MAX_BLOBS_PER_BLOCK * NUMBER_OF_COLUMNS]` | The full data of one-dimensional erasure coding extended blobs (in row major format) |
| `FlatExtendedMatrix` | `List[BLSFieldElement, MAX_BLOBS_PER_BLOCK * FIELD_ELEMENTS_PER_BLOB * NUMBER_OF_COLUMNS]` | The flattened format of `ExtendedMatrix` |

## Configuration

### Data size

| Name | Value | Description |
| - | - | - |
| `NUMBER_OF_COLUMNS` | `uint64((FIELD_ELEMENTS_PER_BLOB * 2) // FIELD_ELEMENTS_PER_CELL)` (= 128) | Number of columns in the extended data matrix. |

### Networking

| Name | Value | Description |
| - | - | - |
| `DATA_COLUMN_SIDECAR_SUBNET_COUNT` | `32` | The number of data column sidecar subnets used in the gossipsub protocol |

### Custody setting

| Name | Value | Description |
| - | - | - |
| `SAMPLES_PER_SLOT` | `8` | Number of `DataColumn` random samples a node queries per slot |
| `CUSTODY_REQUIREMENT` | `1` | Minimum number of subnets an honest node custodies and serves samples from |
| `TARGET_NUMBER_OF_PEERS` | `70` | Suggested minimum peer count |

### Containers

#### `DataColumnSidecar`

```python

> class DataColumnSidecar(Container):

Reviewer 5: I wonder if it's worth sticking with the word `Blob` here for consistency, e.g. `BlobColumnSidecar`, as we previously renamed `data_gas` to `blob_gas` for the reasons described [here](https://github.com/ethereum/EIPs/pull/7354) and #3461 - although this is a different context and calling it `Data` may be acceptable.

    index: ColumnIndex  # Index of column in extended matrix
    column: DataColumn
    kzg_commitments: List[KZGCommitment, MAX_BLOB_COMMITMENTS_PER_BLOCK]
    kzg_proofs: List[KZGProof, MAX_BLOB_COMMITMENTS_PER_BLOCK]
    signed_block_header: SignedBeaconBlockHeader
    kzg_commitments_inclusion_proof: Vector[Bytes32, KZG_COMMITMENTS_INCLUSION_PROOF_DEPTH]
```

### Helper functions

#### `get_custody_columns`

```python

> def get_custody_columns(node_id: NodeID, custody_subnet_count: uint64) -> Sequence[ColumnIndex]:

Reviewer 11: We currently have a function that maps node-id -> subnets and I think we plan to also do node-id -> sync-committees etc. 
@Nashatyrev had an idea called Network Sharding, which just generalizes this concept. 

The idea is we just have a single function that maps node-id -> shard.  
Then we assign attestation subnets, sync-committees and now data columns to shards. 

The benefit here is:
- We have a single function that does the mapping and we don't have to make three different forms of it, making the implementation easier and making it easier to debug
- If we implement it like in: https://github.com/ethereum/consensus-specs/pull/3545, then we make it faster for us to discover peers on columns/subnets/sync committees
- The specs I think might simplify, because this function would just be a mapping from column -> shard.

If others agree, I can draft a spec PR that implements this idea, perhaps it would be more explicit that way.

Reviewer 4: If there are 64 shards but we want to split the blob data in 32 parts, do we map each of the 32 parts to two shards?

Reviewer 11: Yep. The number of shards would be the maximum granularity of splitting nodes. Like we could have 128 shards, in case we wanted to divide the network into 128 and assign subnets etc. I think 64 makes sense at the moment tho, because we currently subscribe to two subnets per node, so 2 shards there also. 

However if we want to have 128 data columns and split this up equally, we might want to have 128 shards, then subnets would be mapped to 4 shards (or vice versa)

Author: > The idea is we just have a single function that maps node-id -> shard

Can it be extended to consider `custody_subnet_count`? i.e., to map node-id to multiple shards.

> assign attestation subnets, sync-committees and now data columns to shards

Would it separate the long-lived subscriptions and the mid-term-lived subscriptions? We may add a [PeerDAS assignment period](https://github.com/ethereum/consensus-specs/blob/87e9702ab0267b4c2320aff31f09a550f744652b/specs/_features/eip7594/das-core.md#subnet-stability) into the deterministic function.

> I can draft a spec PR that implements this idea, perhaps it would be more explicit that way.

I think it would be good to see a PR for it. One minor suggestion: I like the "Network Sharding" naming. However, it might be ambiguous with the data sharding we're mostly talking about.

Reviewer: For reference, the ethresearch post of network sharding is here https://ethresear.ch/t/network-shards-concept-idea/17417

Reviewer 12: > @Nashatyrev had an idea called Network Sharding, which just generalizes this concept.

Here is the concept sketch: https://ethresear.ch/t/network-shards-concept-idea/17417

Reviewer 12: > If others agree, I can draft a spec PR that implements this idea, perhaps it would be more explicit that way.

@AgeManning please feel free to ping me if you want my assistance or maybe want to delegate that to me

Reviewer 11: > Can it be extended to consider custody_subnet_count? i.e., to map node-id to multiple shards.

yep that should be easy to do (If I understand your question). 

> Would it separate the long-lived subscriptions and the mid-term-lived subscriptions? We may add a [PeerDAS assignment period](https://github.com/ethereum/consensus-specs/blob/87e9702ab0267b4c2320aff31f09a550f744652b/specs/_features/eip7594/das-core.md#subnet-stability) into the deterministic function.

This is a good point I'd need to think about. The period of rotation might need to be made a parameter of the function. This would mean that nodes will be on different shards based on this parameter. I guess we shouldn't force rotation periods for different use-cases to be the same. 
In this case, I guess the benefit still is that we have a single function that handles the mappings in a discovery-efficient way.

> @AgeManning please feel free to ping me if you want my assistance or maybe want to delegate that to me

Thanks. I have been traveling and have a conference next week, but I will endeavor to get a PR up for this that we can work from. Feel free to make one earlier if you have the time tho.

Reviewer 11: Apologies for the delay. Here is the proposal: https://github.com/ethereum/consensus-specs/pull/3623

Reviewer 12: > Here is the proposal: #3623

Great! Thanks @AgeManning 

If this PR is adopted then I'd suggest to make `DATA_COLUMN_SIDECAR_SUBNET_COUNT` equal to `NUMBER_OF_COLUMNS`. I.e. one subnet per one column. (that should potentially work well even if number of columns (and thus number of subnets) is of order 2048)

The above would allow push sampling approach: a sampling node could be connected to all (or a subset) of network shards (without subscribing to any column subnets). Some short period (sub-slot) prior to sampling a node subscribes to selected subnets and immediately unsubscribes upon sample receiving. This push approach would allow to 
- avoid extra throughput for sampling: node receives only required samples (columns) 
- almost preserves _unlinkability of queries_: the gossip `subscribe` request could be issued as shortly before expected sample as a second or two 
- a node may always fall back to pull method (req/resp `data_column_sidecars_by_root`) if fails to get a sample via push

    block = signed_block.message
    kzg_commitments_inclusion_proof = compute_merkle_proof(
        block.body,
        get_generalized_index(BeaconBlockBody, 'blob_kzg_commitments'),
    )
    cells_and_proofs = [compute_cells_and_proofs(blob) for blob in blobs]
    blob_count = len(blobs)
    cells = [cells_and_proofs[i][0] for i in range(blob_count)]
    proofs = [cells_and_proofs[i][1] for i in range(blob_count)]
    sidecars = []
    for column_index in range(NUMBER_OF_COLUMNS):
        column = DataColumn([cells[row_index][column_index]
                             for row_index in range(blob_count)])
        kzg_proof_of_column = [proofs[row_index][column_index]
                               for row_index in range(blob_count)]
        sidecars.append(DataColumnSidecar(
            index=column_index,
            column=column,
            kzg_commitments=block.body.blob_kzg_commitments,
            kzg_proofs=kzg_proof_of_column,
            signed_block_header=signed_block_header,
            kzg_commitments_inclusion_proof=kzg_commitments_inclusion_proof,
        ))
    return sidecars
```

## Custody

### Custody requirement

> Each node downloads and custodies a minimum of `CUSTODY_REQUIREMENT` subnets per slot. The particular columns that the node is required to custody are selected pseudo-randomly (more on this below).

Reviewer 3: ```suggestion
Each node downloads and custodies a minimum of `CUSTODY_REQUIREMENT` subnets per slot. The particular subnets that the node is required to custody are selected pseudo-randomly (more on this below).
```
?

> Each node downloads and custodies a minimum of `CUSTODY_REQUIREMENT` subnets per slot. The particular columns that the node is required to custody are selected pseudo-randomly (more on this below).

Reviewer 3: ```suggestion
Each node downloads and custodies a minimum of `CUSTODY_REQUIREMENT` subnets per slot. The particular subnets that the node is required to custody are selected pseudo-randomly (more on this below).
```

> A node *may* choose to custody and serve more than the minimum honesty requirement. Such a node explicitly advertises a number greater than `CUSTODY_REQUIREMENT` via the peer discovery mechanism -- for example, in their ENR (e.g. `custody_lines: 4` if the node custodies `4` subnets each slot) -- up to a `DATA_COLUMN_SIDECAR_SUBNET_COUNT` (i.e. a super-full node).

Reviewer 3: ```suggestion
A node *may* choose to custody and serve more than the minimum honesty requirement. Such a node explicitly advertises a number greater than `CUSTODY_REQUIREMENT` via the peer discovery mechanism -- for example, in their ENR (e.g. `custody_subnets: 4` if the node custodies `4` subnets each slot) -- up to a `DATA_COLUMN_SIDECAR_SUBNET_COUNT` (i.e. a super-full node).
```
could be more clear? lines could still refer to row or column

Author: ```suggestion
A node *may* choose to custody and serve more than the minimum honesty requirement. Such a node explicitly advertises a number greater than `CUSTODY_REQUIREMENT` via the peer discovery mechanism -- for example, in their ENR (e.g. `custody_subnet_count: 4` if the node custodies `4` subnets each slot) -- up to a `DATA_COLUMN_SIDECAR_SUBNET_COUNT` (i.e. a super-full node).
```

> A node stores the custodied columns for the duration of the pruning period and responds to peer requests for samples on those columns.

Reviewer 3: ```suggestion
A node stores the custodied subnets for the duration of the pruning period and responds to peer requests for samples on their respective columns.
```

clarification: so sampling is on columns and custody is on subnets?

### Public, deterministic selection

The particular columns that a node custodies are selected pseudo-randomly as a function (`get_custody_columns`) of the node-id and custody size -- importantly this function can be run by any party as the inputs are all public.

*Note*: increasing the `custody_size` parameter for a given `node_id` extends the returned list (rather than being an entirely new shuffle) such that if `custody_size` is unknown, the default `CUSTODY_REQUIREMENT` will be correct for a subset of the node's custody.

## Peer discovery

> At each slot, a node needs to be able to readily sample from *any* set of columns. To this end, a node should find and maintain a set of diverse and reliable peers that can regularly satisfy their sampling demands.

Reviewer 5: > At each slot, a node needs to be able to readily sample from *any* set of columns.

This includes the subnet that the node is assigned to custody right?

Reviewer 4: Yes. You don't need to do anything when you already have a column you want to sample because you are in that subnet.

Reviewer 5: Got it, thanks! So for nodes that subscribes to all column subnets, they won't be doing any sampling with its peers because they already have all the columns.

Reviewer 5: For nodes custodying at least 50% of the columns, would they still need to sample peers?

Reviewer: > For nodes custodying at least 50% of the columns, would they still need to sample peers?

I think they shouldn't because the data is now surely available.

## Peer scoring

Due to the deterministic custody functions, a node knows exactly what a peer should be able to respond to. In the event that a peer does not respond to samples of their custodied rows/columns, a node may downscore or disconnect from a peer.

## DAS providers

A DAS provider is a consistently-available-for-DAS-queries, super-full (or high capacity) node. To the p2p, these look just like other nodes but with high advertised capacity, and they should generally be able to be latently found via normal discovery.

DAS providers can also be found out-of-band and configured into a node to connect to directly and prioritize. Nodes can add some set of these to their local configuration for persistent connection to bolster their DAS quality of service.

Such direct peering utilizes a feature supported out of the box today on all nodes and can complement (and reduce attackability and increase quality-of-service) alternative peer discovery mechanisms.

## A note on fork choice

> *Fork choice spec TBD, but it will just be a replacement of `is_data_available()` call in Deneb with column sampling instead of full download. Note the `is_data_available(slot_N)` will likely do a `-1` follow distance so that you just need to check the availability of slot `N-1` for slot `N` (starting with the block proposer of `N`).*

Reviewer 8: Is there more information on how the trailing slot fork choice works during network partitions or longer forks ? During any event where you might temporarily have your peers with split views(hard fork, consensus bug), the peers that you are connected to for the particular column subnets will neither have the block or the data. So when sampling for `n-1` from these peers, they will return nothing and you would mark the data as unavailable. 

How does sampling work when the network is undergoing stress and possible churn ? ( constant peer disconnections and new connections to find peers with the desired column subnets) . Is there a mode where we can fall back to so that a block and its data which was correctly broadcasted and propagated is marked as available rather than unavailable during such a situation.

    - [`get_custody_columns`](#get_custody_columns)
    - [`compute_extended_matrix`](#compute_extended_matrix)
    - [`recover_matrix`](#recover_matrix)
    - [`get_data_column_sidecars`](#get_data_column_sidecars)
- [Custody](#custody)
  - [Custody requirement](#custody-requirement)
  - [Public, deterministic selection](#public-deterministic-selection)
- [Peer discovery](#peer-discovery)
- [Extended data](#extended-data)
- [Column gossip](#column-gossip)
  - [Parameters](#parameters)
- [Peer sampling](#peer-sampling)
- [Peer scoring](#peer-scoring)
- [Reconstruction and cross-seeding](#reconstruction-and-cross-seeding)
- [DAS providers](#das-providers)
- [A note on fork choice](#a-note-on-fork-choice)
- [FAQs](#faqs)
  - [Row (blob) custody](#row-blob-custody)
  - [Subnet stability](#subnet-stability)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->
<!-- /TOC -->

## Custom types

We define the following Python custom types for type hinting and readability:

| Name | SSZ equivalent | Description |
| - | - | - |
| `DataColumn` | `List[Cell, MAX_BLOB_COMMITMENTS_PER_BLOCK]` | The data of each column in EIP-7594 |

> | `ExtendedMatrix` | `List[Cell, MAX_BLOBS_PER_BLOCK * NUMBER_OF_COLUMNS]` | The full data of one-dimensional erasure coding extended blobs (in row major format) |

Reviewer 9: Should `ExtendedMatrix` use `MAX_BLOB_COMMITMENTS_PER_BLOCK` too?

## Configuration

### Data size

| Name | Value | Description |
| - | - | - |

> | `NUMBER_OF_COLUMNS` | `uint64((FIELD_ELEMENTS_PER_BLOB * 2) // FIELD_ELEMENTS_PER_CELL)` (= 128) | Number of columns in the extended data matrix. |

Reviewer 13: ```suggestion
| `NUMBER_OF_COLUMNS` | `uint64(FIELD_ELEMENTS_PER_EXT_BLOB // FIELD_ELEMENTS_PER_CELL)` (= 128) | Number of columns in the extended data matrix. |
```

Low priority, feel free to merge without this.

### Networking

| Name | Value | Description |
| - | - | - |
| `DATA_COLUMN_SIDECAR_SUBNET_COUNT` | `32` | The number of data column sidecar subnets used in the gossipsub protocol |

### Custody setting

| Name | Value | Description |
| - | - | - |
| `SAMPLES_PER_SLOT` | `8` | Number of `DataColumn` random samples a node queries per slot |

> | `CUSTODY_REQUIREMENT` | `1` | Minimum number of subnets an honest node custodies and serves samples from |

Reviewer 4: As discussed with @dankrad, I think this should be higher than 1, because it is the most important parameter in determining how strong are the pre-sampling guarantees we get, i.e., how hard it is for an unavailable block to get a majority of the votes and become canonical, even when employing the trailing fork-choice. Ideally, we would want to ensure that only a small fraction of the validators can be tricked into voting for an unavailable block.

Imo, a suitable value would be >= 6. We might then also want to increase the number of subnets to 64, to keep the bandwidth consumption low. For reference, `(CUSTODY_REQUIREMENT = 6, DATA_COLUMN_SIDECAR_SUBNET_COUNT = 64)` and everything else unchanged means that the total bandwidth consumption with `MAX_BLOBS_PER_BLOCK = 16` is 60% of 4844. With 32 blobs, it's 120% of 4844. It can also be lowered by increasing the subnet count further if needed.

Reviewer 4: I agree that the name should be `SUBNET_CUSTODY_REQUIREMENT`. Still, custodying a single subnet is not the same as custodying 4 *random* columns, and in fact gives you almost no pre-sampling guarantees.

For example, someone can make 49% *of the subnets* available (not making a subnet available here means not publishing any of the columns for that subnet), and 49% of validators will still succeed in downloading all columns they need. If on the other end the custody requirement had been 4 random columns, making 49% of columns available would only trick 1/16 of the validators. This is also the case if validators custody 4 subnets.

| `TARGET_NUMBER_OF_PEERS` | `70` | Suggested minimum peer count |

### Containers

#### `DataColumnSidecar`

```python
class DataColumnSidecar(Container):
    index: ColumnIndex  # Index of column in extended matrix
    column: DataColumn
    kzg_commitments: List[KZGCommitment, MAX_BLOB_COMMITMENTS_PER_BLOCK]
    kzg_proofs: List[KZGProof, MAX_BLOB_COMMITMENTS_PER_BLOCK]
    signed_block_header: SignedBeaconBlockHeader
    kzg_commitments_inclusion_proof: Vector[Bytes32, KZG_COMMITMENTS_INCLUSION_PROOF_DEPTH]
```

### Helper functions

#### `get_custody_columns`

```python
def get_custody_columns(node_id: NodeID, custody_subnet_count: uint64) -> Sequence[ColumnIndex]:
    assert custody_subnet_count <= DATA_COLUMN_SIDECAR_SUBNET_COUNT

    subnet_ids = []
    i = 0
    while len(subnet_ids) < custody_subnet_count:
        subnet_id = (

>             bytes_to_uint64(hash(uint_to_bytes(uint64(node_id + i)))[0:8])

Reviewer: > Why cast this down to 64-bit? why not just pass in the 256-bit value to `hash`?

I think it's because you need to do the `+ i` thing. You need to cast it to 64-bit first to do the math. I guess doing the math under `uint256` is not supported by all the programming languages.

However, I think it should be `uint64(node_id) + i` rather than `uint64(node_id + i)`

            % DATA_COLUMN_SIDECAR_SUBNET_COUNT
        )
        if subnet_id not in subnet_ids:
            subnet_ids.append(subnet_id)
        i += 1

>     assert len(subnet_ids) == len(set(subnet_ids))

Reviewer: Agree with this. This assertion is always false. The intention is only to catch a bug.

    )
    cells_and_proofs = [compute_cells_and_proofs(blob) for blob in blobs]
    blob_count = len(blobs)
    cells = [cells_and_proofs[i][0] for i in range(blob_count)]
    proofs = [cells_and_proofs[i][1] for i in range(blob_count)]
    sidecars = []
    for column_index in range(NUMBER_OF_COLUMNS):
        column = DataColumn([cells[row_index][column_index]
                             for row_index in range(blob_count)])
        kzg_proof_of_column = [proofs[row_index][column_index]
                               for row_index in range(blob_count)]
        sidecars.append(DataColumnSidecar(
            index=column_index,
            column=column,
            kzg_commitments=block.body.blob_kzg_commitments,
            kzg_proofs=kzg_proof_of_column,
            signed_block_header=signed_block_header,
            kzg_commitments_inclusion_proof=kzg_commitments_inclusion_proof,
        ))
    return sidecars
```

## Custody

### Custody requirement

Each node downloads and custodies a minimum of `CUSTODY_REQUIREMENT` subnets per slot. The particular subnets that the node is required to custody are selected pseudo-randomly (more on this below).

> A node *may* choose to custody and serve more than the minimum honesty requirement. Such a node explicitly advertises a number greater than `CUSTODY_REQUIREMENT` via the peer discovery mechanism -- for example, in their ENR (e.g. `custody_subnet_count: 4` if the node custodies `4` subnets each slot) -- up to a `DATA_COLUMN_SIDECAR_SUBNET_COUNT` (i.e. a super-full node).

Reviewer 5: Implementation question - my apologies if it's off-topic for this PR - is there any value for a super-full node to custody the extended columns / rows (besides being able to serve data efficiently)? as it would always be able to reconstruct the full matrix without asking from its peers. 

If it stores just the raw blobs, it would only need to store 50% of the samples for 1D extended blobs, or just 25% of the samples of the 2D extended blobs:

| Max Blobs \ Data Stored | Storing raw blob only | 1D extended | 2D extended |
| --------- | -------- | ----------- | ----------- |
| 6         | 96 GB    | 192 GB      | 384 GB      |
| 32        | 512 GB   | 1 TB        | 2 TB        |
| 64        | 1TB      | 2 TB        | 4 TB        |
| 128       | 2TB      | 4 TB        | 8 TB        |

Reviewer 7: > Implementation question - my apologies if it's off-topic for this PR - is there any value for a super-full node to custody the extended columns / rows (besides being able to serve data efficiently)? as it would always be able to reconstruct the full matrix without asking from its peers.

Reconstructing a column when you have only row encoding is quite expensive. Say you need column `j`. Then for every segment `i,j` in row `i`, you would need to read half of the row `i` to do the reconstruction. This basically means reading in the whole data amount to reconstruct a single row. 

So while you are always able to reconstruct, it can be costly, reading in and doing erasure coding on lots of data. I think it should be an implementation decision, but for the 1D extension I would keep the whole extended data for recent blocks, and only maybe discard half for old ones.

For the 2D extension, I would most probably store half of it, but in a more intelligent way than keeping the K columns or K rows. Basically what you want to do is to **keep the upper left corner** (original data in the systematic encoding), **and the lower right corner** ("double-parity" data) of the encoding. This allows you to efficiently restore any row, column or individual segment, while keeping only half of the data. Again, for old data that most probably won't be queried, you might want to fall back to keeping only the original (or some other K half-rows), while paying the price of a full restore when queried.

We might even go further, and instead of keeping the upper left corner and lower right corner, select a set `A` of K row IDs and a set `B` of K column IDs. Then you store chunk `i,j` iff
$i \in A \land j \in B \lor i \notin A \land j \notin B$
This allows the same fast reconstruction, but randomises who stores what. This can be done with local randomness to improve resilience, or with peerID based randomness to reduce load of queries from good behaving nodes who can query nodes where they know the node can serve cheap without reconstruction.

Of course, a super-full node would have to balance the above with its own needs to access the original data. If you actually want to read the original data frequently for your own use, it would be stupid not to store the original part (upper left corner) of the systematic encoding.

Reviewer 14: > Implementation question - my apologies if it's off-topic for this PR - is there any value for a super-full node to custody the extended columns / rows (besides being able to serve data efficiently)? as it would always be able to reconstruct the full matrix without asking from its peers.
> 
> If it stores just the raw blobs, it would only need to store 50% of the samples for 1D extended blobs, or just 25% of the samples of the 2D extended blobs:
> Max Blobs \ Data Stored 	Storing raw blob only 	1D extended 	2D extended
> 6 	96 GB 	192 GB 	384 GB
> 32 	512 GB 	1 TB 	2 TB
> 64 	1TB 	2 TB 	4 TB
> 128 	2TB 	4 TB 	8 TB

All these are off by a factor of 2, assuming that target blobs are 1/2 of max blobs. (The EIP-1559 style pricing mechanism makes it impossible to significantly exceed the target for such a long period by increasing the price exponentially)

*Note*: while high-capacity and super-full nodes are high value with respect to satisfying sampling requirements, a node should maintain a distribution across node capacities as to not centralize the p2p graph too much (in the extreme becomes hub/spoke) and to distribute sampling load better across all nodes.

*Note*: A DHT-based peer discovery mechanism is expected to be utilized in the above. The beacon-chain network currently utilizes discv5 in a similar method as described for finding peers of particular distributions of attestation subnets. Additional peer discovery methods are valuable to integrate (e.g., latent peer discovery via libp2p gossipsub) to add a defense in breadth against one of the discovery methods being attacked.

## Extended data

In this construction, we extend the blobs using a one-dimensional erasure coding extension. The matrix comprises maximum `MAX_BLOBS_PER_BLOCK` rows and fixed `NUMBER_OF_COLUMNS` columns, with each row containing a `Blob` and its corresponding extension. `compute_extended_matrix` demonstrates the relationship between blobs and custom type `ExtendedMatrix`.

## Column gossip

### Parameters

> For each column -- use `data_column_sidecar_{subnet_id}` subnets, where each column index maps to the `subnet_id`. The sidecars can be computed with `get_data_column_sidecars(signed_block: SignedBeaconBlock, blobs: Sequence[Blob])` helper.

Reviewer 5: ```suggestion
For each column -- use `data_column_sidecar_{subnet_id}` subnets, where `subnet_id` can be computed with the `compute_subnet_for_data_column_sidecar(column_index: ColumnIndex)` helper. The sidecars can be computed with the `get_data_column_sidecars(signed_block: SignedBeaconBlock, blobs: Sequence[Blob])` helper.
```
